nearcore/src/cold_storage.rs

```diff
@@ -0,0 +1,183 @@
+fn cold_store_copy(
+    hot_store: &Store,
+    cold_store: &Store,
+    cold_db: &Arc<ColdDB>,
+    genesis_height: BlockHeight,
+    runtime: &Arc<NightshadeRuntime>,
+) -> anyhow::Result<ColdStoreLoopStatus> {
+    let cold_head = cold_store.get_ser::<Tip>(DBCol::BlockMisc, HEAD_KEY)?;
+    let cold_head_height = match cold_head {
+        Some(cold_head) => cold_head.height,
+        None => genesis_height,
+    };
+
+    // If FINAL_HEAD is not set for hot storage though, we default it to 0.
+    let hot_final_head = hot_store.get_ser::<Tip>(DBCol::BlockMisc, FINAL_HEAD_KEY)?;
+    let hot_final_head_height = match hot_final_head {
+        Some(hot_final_head) => hot_final_head.height,
+        None => genesis_height,
+    };
+
+    tracing::debug!(target: "cold_store", "cold store loop, cold_head {}, hot_final_head {}", cold_head_height, hot_final_head_height);
+
+    if cold_head_height > hot_final_head_height {
+        tracing::error!(target: "cold_store", "Error, cold head is ahead of final head, this should never take place!");
+    }
+
+    if cold_head_height >= hot_final_head_height {
+        return Ok(ColdStoreLoopStatus::Sleep);
+    }
+
+    let next_height = cold_head_height + 1;
+
+    // Here it should be sufficient to just read from hot storage.
+    // Because BlockHeight is never garbage collectable and is not even copied to cold.
+    let cold_head_hash =
+        hot_store.get_ser::<CryptoHash>(DBCol::BlockHeight, &cold_head_height.to_le_bytes())?;
+    let cold_head_hash = match cold_head_hash {
+        Some(cold_head_hash) => cold_head_hash,
+        None => return Err(anyhow::anyhow!("Failed to read the cold head hash")),
+    };
+
+    // The previous block is the cold head so we can use it to get epoch id.
+    let epoch_id = &runtime.get_epoch_id_from_prev_block(&cold_head_hash)?;
+    let shard_layout = runtime.get_shard_layout(epoch_id)?;
+
+    update_cold_db(cold_db, hot_store, &shard_layout, &next_height)?;
+
+    update_cold_head(cold_db, hot_store, &next_height)?;
+
+    if next_height >= hot_final_head_height {
+        Ok(ColdStoreLoopStatus::Sleep)
+    } else {
+        Ok(ColdStoreLoopStatus::Continue)
+    }
+}
+
+// This method will copy data from hot storage to cold storage in a loop.
+// It will try to copy blocks as fast as possible up until cold head = final head.
+// Once the cold head reaches the final head it will sleep for one second before
+// trying to copy data at the next height.
+// TODO clean up the interface, currently we need to pass hot store, cold store and
+// cold_db which is redundant.
+fn cold_store_loop(
+    keep_going: Arc<AtomicBool>,
+    hot_store: Store,
+    cold_store: Store,
+    cold_db: Arc<ColdDB>,
+    genesis_height: BlockHeight,
+    runtime: Arc<NightshadeRuntime>,
+) {
+    tracing::info!(target: "cold_store", "starting cold store loop");
+
+    // TODO - is there a better way to schedule a function to run periodically?
```

Comment:
you should ask on rust channel in zulip.

Reply:
will do!

Reply:
The early suggestions are to consider actix or a dedicated thread (the current solution). The rocks db api is not async which makes it not suitable for async solution such as actix or tokio because despite being IO it wouldn't give the thread back to the threadpool.